noir-projects/noir-contracts/contracts/token_contract/src/main.nr

```diff
@@ -613,64 +616,44 @@ contract Token {
         finalization_payload.emit();
     }
 
-    /// We need to use different randomness for the user and for the fee payer notes because if the randomness values
-    /// were the same we could fingerprint the user by doing the following:
-    ///      1) randomness_influence = fee_payer_point - G_npk * fee_payer_npk =
-    ///                              = (G_npk * fee_payer_npk + G_rnd * randomness + G_slot * fee_payer_slot)
-    ///                                - G_npk * fee_payer_npk - G_slot * fee_payer_slot =
-    ///                              = G_rnd * randomness
-    ///      2) user_fingerprint = user_point - randomness_influence =
-    ///                          = (G_npk * user_npk + G_rnd * randomness + G_slot * user_slot) - G_rnd * randomness =
-    ///                          = G_npk * user_npk + G_slot * user_slot
-    ///      3) Then the second time the user would use this fee paying contract we would recover the same fingerprint
-    ///         and link that the 2 transactions were made by the same user. Given that it's expected that only
-    ///         a limited set of fee paying contracts will be used and they will be known, searching for fingerprints
-    ///         by trying different fee payers is a feasible attack.
-    ///
-    /// Note 1: fee_payer_npk is part of the fee_payer address preimage derivation, and is assumed to be known. So
-    //          if we have a known set of fee payer contract addresses getting fee_payer_npk and fee_payer_slot is
-    //          trivial (slot is derived in a `Map<...>` as a hash of balances map slot and a fee payer address).
-    /// Note 2: fee_payer_point and user_point above are public information because they are passed as args to
-    ///         the public `complete_refund(...)` function.
     // docs:start:setup_refund
+    /// Called by fee payer contract (FPC) to set up a refund for a user during the private fee payment flow.
     #[private]
     fn setup_refund(
-        fee_payer: AztecAddress, // Address of the entity which will receive the fee note.
         user: AztecAddress, // A user for which we are setting up the fee refund.
-        funded_amount: Field, // The amount the user funded the fee payer with (represents fee limit).
+        max_fee: Field, // The maximum fee a user is willing to pay for the tx (denominated in FPC's accepted asset)
```

Comment:
> denominated in FPC's accepted asset

The denomination is already clear within this token contract, since this token contract _is_ the denomination.
Ideally, a token contract shouldn't mention "FPCs" or "fees".

Reply:
I'll spend some time pondering how to abstract "fee logic" from the token contract, without sacrificing efficiency. The token contract is kind of ugly with all this included, right?